File: quick_sort/quick_sort.py
def swap(lst, index1, index2):
    lst[index1], lst[index2] = lst[index2], lst[index1]

def pivot(lst, pivot_index, end_index):
    swap_index = pivot_index

    for i in range(pivot_index + 1, end_index + 1):
        if lst[i] < lst[pivot_index]:
            swap_index += 1
            swap(lst, swap_index, i)

    swap(lst, pivot_index, swap_index)
    return swap_index
    
def quick_sort_helper(lst, left, right):
    if left < right:
        pivot_index = pivot(lst, left, right)
        quick_sort_helper(lst, left, pivot_index - 1)
        quick_sort_helper(lst, pivot_index + 1, right)
    return lst

def quick_sort(lst):
    return quick_sort_helper(lst, 0, len(lst) - 1)

print(quick_sort([1, 2, 7, 8, 3, 10, 11, 4, 9, 5, 6]))

File: quick_sort/quick_sort.go
package main

import "fmt"

func swap(arr []int, i, j int) {
	arr[i], arr[j] = arr[j], arr[i]
}

func pivot(arr []int, pivotIndex, endIndex int) int {
	swapIndex := pivotIndex

	for i := pivotIndex + 1; i <= endIndex; i++ {
		if arr[i] < arr[pivotIndex] {
			swapIndex++
			swap(arr, swapIndex, i)
		}
	}
	swap(arr, pivotIndex, swapIndex)
	return swapIndex
}

func quickSortHelper(arr []int, left, right int) {
	if left < right {
		pivotIndex := pivot(arr, left, right)
		quickSortHelper(arr, left, pivotIndex-1)
		quickSortHelper(arr, pivotIndex+1, right)
	}
}

func quickSort(arr []int) {
	quickSortHelper(arr, 0, len(arr)-1)
}

func main() {
	arr := []int{1, 2, 7, 8, 3, 10, 11, 4, 9, 5, 6}
	quickSort(arr)
	fmt.Println(arr)
}
